[3] Allow to set password empty passwords and add database password back

HLeithner opened the pull request:
Pull Request for Issue #31671 and #31452 .

### Summary of Changes
Add an attribute `lock` to the password field which does the following 3 things:

* set the `disabled` attribute to the password field (this means the field will not be send to the server)
* add a button with the label "Modify" to the password field (this toggles the disable attribute of the password field)
* removes the value from the password field

Also the database password field is back, since we have a database connection checker we can be safe that we don't break the installation on save.

### Testing Instructions
* Change change passwords like database password, smtp password, ftp password and others in the global configurations.
* set the password to an empty value
* Test other password fields like user passwords

### Actual result BEFORE applying this Pull Request
* Changing passwords work
* Setting an empty password doesn't work
* Other Password fields at the global configuration work

### Expected result AFTER applying this Pull Request
* Changing passwords work
* *Setting an empty password will set the password to an empty value now*
* Other Password fields at the global configuration still work


### Documentation Changes Required
https://docs.joomla.org/Password_form_field_type

HLeithner committed:
Allow to set password empty passwords and add database password back

brianteeman commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -58,6 +58,8 @@ COM_CONFIG_FIELD_DATABASE_HOST_DESC="The hostname for your database entered duri
 COM_CONFIG_FIELD_DATABASE_HOST_LABEL="Host"
 COM_CONFIG_FIELD_DATABASE_NAME_DESC="The name for your database entered during the installation process. Do not edit this field unless absolutely necessary (eg the transfer of the database to a new hosting provider)."
 COM_CONFIG_FIELD_DATABASE_NAME_LABEL="Database Name"
+COM_CONFIG_FIELD_DATABASE_PASSWORD_DESC="The password for access to your database entered during the installation process. Do not edit this field unless absolutely necessary (eg the transfer of the database to a new hosting provider)."
```
> The password for access to your database entered during the installation process

Obviously that makes sense the first time someone reads it. But once they have changed it once it is no longer a true statement

HLeithner replied:
It's copy paste from the username so this would be wrong too... Can you make a suggestion please?

brianteeman replied:
COM_CONFIG_FIELD_DATABASE_PASSWORD_DESC="The password for access to your database. Do not edit this field unless absolutely necessary (eg after the transfer of the database to a new hosting provider)."

brianteeman commented on layouts/joomla/form/field/password.php:
```diff
@@ -42,6 +42,7 @@
  * @var   array    $options         Options available for this field.
  * @var   array    $inputType       Options available for this field.
  * @var   string   $accept          File types that are accepted.
+ * @var   boolean  $lock            File types that are accepted.
```
file types ??

HLeithner replied:
Thanks will fix it

brianteeman commented:
The parts I have tested (global config db password and user password) work correctly

I do not understand why the lock/modify button is present

There must be a clear feedback that the button has been clicked and the input is now disabled. The subtle background change on the input is not sufficient

### Locked
![image](https://user-images.githubusercontent.com/1296369/102283218-db46ee80-3f29-11eb-97bd-c1650914ec6a.png)

### Unlocked
![image](https://user-images.githubusercontent.com/1296369/102283248-eef25500-3f29-11eb-9ef6-32f18c5ddef9.png)

HLeithner commented:
Don't know how to make it more clear because that's how a disabled field looks like

brianteeman commented:
> Don't know how to make it more clear because that's how a disabled field looks like

By changing the button. 

Look at the password login field where it changes from an open to a closed eye - although in this case I would try something like using a padlock in open and closed state and then also toggle the text between "protected" and "modify" or something like that

brianteeman commented:
I assume that you realise that as soon as you change the database you will be immediately logged out as the default sessionhandler is database

HLeithner commented:
> I assume that you realise that as soon as you change the database you will be immediately logged out as the default sessionhandler is database

Yes but that's true for all database settings.

brianteeman commented:
Was just checking. Personally that's a great reason for all of these database fields to be readonly as I cant think of a real world scenario where I have a working site that I will ever want to change the database of. In the extremely rare scenario that someone might come up with you can always edit the configuration.php.

HLeithner commented:
I use them regularly while doing joomla updates and I see no problem doing this also because we have a protection against miss configuration.

On the other side you can can break your site with many other settings in the global configuration. But at the moment I want to fix the unset problem so I can build the rc....

brianteeman commented on layouts/joomla/form/field/password.php:
```diff
@@ -80,8 +96,13 @@
 );
 
 ?>
+<?php if ($lock): ?><span class="input-append"><?php endif; ?>
 <input type="password" name="<?php
 echo $name; ?>" id="<?php
 echo $id; ?>" value="<?php
 echo htmlspecialchars($value, ENT_COMPAT, 'UTF-8'); ?>" <?php
 echo implode(' ', $attributes); ?> />
+<?php if ($lock): ?>
+<button type="button" id="<?php echo $id; ?>_lock" class="btn btn-info" data-toggle="button"><?php echo JText::_('JMODIFY'); ?></button>
+</span>
+<?php endif; ?>
```
```suggestion
<?php if ($lock): ?>
    <span class="input-append">
<?php endif; ?>

<input 
    type="password" 
    name="<?php echo $name; ?>" 
    id="<?php echo $id; ?>" 
    value="<?php echo htmlspecialchars($value, ENT_COMPAT, 'UTF-8'); ?>" 
    <?php echo implode(' ', $attributes); ?>
/>
<?php if ($lock): ?>
    <button type="button" id="<?php echo $id; ?>_lock" class="btn btn-info" data-toggle="button"><?php echo JText::_('JMODIFY'); ?></button>
    </span>
<?php endif; ?>
```

HLeithner committed:
Update layouts/joomla/form/field/password.php

Co-authored-by: Brian Teeman <brian@teeman.net>

wilsonge committed:
Brace on new line

wilsonge commented:
I think this works well for the database field. My only comment would be that I think the fields that are hidden by showon (e.g. ftp/proxy) need to be in edit mode by default when first enabled somehow :/ Because otherwise they look like they contain values when they are actually empty and will probably confuse our users. Example field with no value

<img width="489" alt="Screenshot 2020-12-16 at 00 23 33" src="https://user-images.githubusercontent.com/1986000/102288706-f0754a80-3f34-11eb-84c9-3d7dd7b9e631.png">

HLeithner commented:
> My only comment would be that I think the fields that are hidden by showon (e.g. ftp/proxy) need to be in edit mode by default when first enabled somehow :/ Because otherwise they look like they contain values when they are actually empty and will probably confuse our users. Example field with no value
> 
> <img alt="Screenshot 2020-12-16 at 00 23 33" width="489" src="https://user-images.githubusercontent.com/1986000/102288706-f0754a80-3f34-11eb-84c9-3d7dd7b9e631.png">
> 
> Especially as we're using the string modify which kinda implies something already exists.

I would remove the hint completely

HLeithner committed:
Remove hint and update button text

wilsonge commented:
Also I think if you enter a default value for the password field then currently the value is set into the session and hitting save subsequently as the password field is disabled constantly uses the value from the session until you either hit cancel or remodify it. However it's not clear from the UI (obviously) that you're using the session value as opposed to the original value of the field because the field switches back to the disabled mode on the page reload.

HLeithner commented:
Now it looks like this in protected mode
![image](https://user-images.githubusercontent.com/1497730/102289602-5ff03780-3f3f-11eb-98a9-805622c26672.png)
and like this in unprotected mode
![image](https://user-images.githubusercontent.com/1497730/102289655-7b5b4280-3f3f-11eb-897f-ab10317508fd.png)

HLeithner commented:
as alternative a one-way "modify" button without return is possible too

wilsonge commented:
> as alternative a one-way "modify" button without return is possible too

I think this would make more sense. Combined with automatic unlock if the field has been submitted with an invalid value preventing save (i.e. the password value is stuck in the session as opposed to the configuration.php value)

brianteeman commented:
Sorry I'm not commenting on accessibility. Ask @carcam and the accessibility team. It's a big enough team that I'm sure one of them could find the time to respond.

SniperSister commented:
> I would remove the hint completely

Can't we add the hint conditionally? If a non-empty value is set, add the hint; if the value is empty, leave the field blank. That way we have a clear indicator for the user.

HLeithner committed:
Set hint to password length, unset password on lock

HLeithner commented:
@wilsonge 
> > as alternative a one-way "modify" button without return is possible too
> 
> I think this would make more sense to me - although no sure if the absence of the button would constitute valid a11y - brianteeman any feedback on that? Combined with automatic unlock if the field has been submitted with an invalid value preventing save (i.e. the password value is stuck in the session as opposed to the configuration.php value)

I see no clean way to address this, yes the password is saved in the session but will not be set in the field again because it's replaced by an empty string.

@SniperSister 
> > I would remove the hint completely
> 
> Can't we add the hint conditionally? If a non-empty value is set, add the hint; if the value is empty, leave the field blank. That way we have a clear indicator for the user.

I renamed the button to "Previous" after pressing "Modify" and added the hint back with the count of characters of the original password. Also if you press "Previous" the password field is set to an empty string again and the hint (placeholder) is shown again.

chmst commented:
I have tested the functions and it works well. 

my2cent for the buttons: I suggest "edit" (if it is disabled) and "protect" (when enabled) instead of "modify" and "protected".

For a11y the  button could have a title-attribute "open the input field for edit"  and "protect the input field" for example.

HLeithner commented:
> my2cent for the buttons: I suggest "edit" (if it is disabled) and "protect" (when enabled) instead of "modify" and "protected".

"edit" is fine for me, I changed "protected" to "Previous" could be "cancel" too

> For a11y the button could have a title-attribute "open the input field for edit" and "protect the input field" for example.

This is a title correct? So a description like "Allow to change the password" and "Cancel the password change" or so is too much?

HLeithner committed:
Change disable button label to JCANCEL

Quy commented on administrator/language/en-GB/en-GB.com_config.ini:
```diff
@@ -58,6 +58,8 @@ COM_CONFIG_FIELD_DATABASE_HOST_DESC="The hostname for your database entered duri
 COM_CONFIG_FIELD_DATABASE_HOST_LABEL="Host"
 COM_CONFIG_FIELD_DATABASE_NAME_DESC="The name for your database entered during the installation process. Do not edit this field unless absolutely necessary (eg the transfer of the database to a new hosting provider)."
 COM_CONFIG_FIELD_DATABASE_NAME_LABEL="Database Name"
+COM_CONFIG_FIELD_DATABASE_PASSWORD_DESC="The password for access to your database. Do not edit this field unless absolutely necessary (eg after the transfer of the database to a new hosting provider)"
```
```suggestion
COM_CONFIG_FIELD_DATABASE_PASSWORD_DESC="The password for access to your database. Do not edit this field unless absolutely necessary (eg after the transfer of the database to a new hosting provider)."
```

bembelimen committed:
Update administrator/language/en-GB/en-GB.com_config.ini

Co-authored-by: Quy <quy@fluxbb.org>

wilsonge commented:
I have tested this item :white_check_mark: successfully on d3698cd7becae9b43056126ffe32db41e21b11ae<br /><br />I think this is good enough to ship. Definitely can improve it further. But fixes the "regression" in functionality for the next release. I'd consider at least being able to change SMTP credentials between local dev and prod sites essential (when I'm restoring via akeeba or whatever).<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31672">issues.joomla.org/tracker/joomla-cms/31672</a>.</sub>

webgras commented:
I have tested this item :white_check_mark: successfully on d3698cd7becae9b43056126ffe32db41e21b11ae<br /><br />I tested password fields (FTP, SMTP and DB in global configuration - could not find others) to set a new one, change it and empty it.
I tested also the normal user registration in frontend and creating a user in the backend.
I changed passwords successfully in user management and in account details.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31672">issues.joomla.org/tracker/joomla-cms/31672</a>.</sub>

HLeithner committed:
cs

PhilETaylor commented:
Ping. you might want to fix this bug... https://github.com/joomla/joomla-cms/issues/31716

HLeithner commented:
I already see you report and try to reproduce it and of course I will fix it
